### DISCUSSION TITLE
-DBOARD_ROOT (or its alternative via CMakeLists.txt) issues

### DISCUSSION THREAD

**Discussion by rjmaris** [zephyrproject-rtos/zephyr#58900] [Q&A] [answered]
A documentation clarity question and a usage question is stated here.

**Documentation**: [here](https://docs.zephyrproject.org/latest/develop/application/index.html?highlight=dboard_root#boards), to my opinion, clarity can be improved. I interpreted `-DBOARD_ROOT` as definition of the (normally) relative path to the boards directory. Since this is `boards` in my case, I defined `-DBOARD_ROOT=boards`. Far later I discovered that I had to define it as `-DBOARD_ROOT=./`. It was after the build stopped for the x_th time with error: when I decided to scroll back to above the long(!) list of boards that is presented above the build process output. Then I saw the brown warning with the interpreted full path - with an extra `boards` on its end. Would it make sense to change the wording instead of `<path to boards>` to `<path where the boards directory is located>`?

**Usage**: I had an extra difficulty when I specified e.g. `-b samples/net/telnet` instead of the code within the application folder. I'd consider this a useful use case when I have my own board folder out-of-tree: allowing sample testing with own board definition.
The problem: As soon as another path is specified, the `BOARD_ROOT=./`is then related to the selected sample code directory. Hence I'm forced to specify a full path (even ~/appfolder/ is not considered a full path). How can I manage it using an entry in `CMakeLists.txt`? I cannot use `CMAKE_CURRENT_SOURCE` or `CMAKE_CURRENT_LIST_DIR`. I'd prefer a solution where I don't need to change an entry regardless the location of the code to be built. And regardless whether I once move my out -of-tree application to another path.

**Comment by DerekSnell** [comment, marked as answer]
Hi @rjmaris ,

Regarding the documentation, your proposal sounds good to me.  But I suggest you submit a PR with this change.  That will get the attention of the maintainers and others interested, and make it easy to update the doc and merge it.

As for the relative paths from the application folder, I have run into similar challenges.  I have an example below of `CmakeLists.txt` that I am using in a T2 Star West topology.  In this case, I am not using a custom board folder like you.  But I am including a static library in my build that is not located with the application, and this avoids using a relative path from the `CMAKE_CURRENT_SOURCE_DIR`.  Since West installs this module in a related path to the `zephyr` folder, I use the `ZEPHYR_BASE` environment variable in my library path.  Maybe that will be an option for you.  Best regards

```
find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})

set(LIB_DIR  $ENV{ZEPHYR_BASE}/../modules/myLibFolder)
add_library(myLib STATIC IMPORTED GLOBAL)
set_target_properties(myLib PROPERTIES IMPORTED_LOCATION ${LIB_DIR}/mylib.a)
target_link_libraries(app PUBLIC myLib)
include_directories(app ${LIB_DIR})

FILE(GLOB app_sources src/*.c)
target_sources(app PRIVATE ${app_sources})
```

**Reply by rjmaris** [reply]
Hi Derek, regarding PR you mean Pull request? I'll write an issue, accompanied by a patch file (I have no forked repository copy, only a local cloned copy of the repository).

Regarding the topic itself: I'd prefer to have the board files within a subdirectory of the application, even when I'd later add another set of board files, e.g. upon migration of the application to another architecture later on. However, your solution makes sense when a custom board would be used with alternative applications as well. For that reason I'll reconsider my current resouces storage structure.
BTW: Meanwhile I also got the path problem when I have applied overlay files: When I (temporarily) build a demo project from the zephyr/samples tree, it uses overlay files from there - when I don't supply the path for overlays in `CMakeLists.txt` or as a `west build` argument...

**Comment by DerekSnell** [comment]
Hi @rjmaris ,
Yes, I meant a Pull Request.  You are welcome to create an issue.  But submitting a PR typically resolves the issue faster.  With an Issue, you are relying on others to solve the problem.  And sharing a patch file in an Issue is OK.  But with a PR, reviewers can easily see the proposed changes and add their feedback per line.  This specific documentation change you pointed out would be a very simple change, and a great PR to get started with to learn the PR process.

It is also very easy to use GitHub to create a fork of the Zephyr repo in your GitHub account.  Then you create a branch in your fork, make the change, push the branch to your GitHub repo, and use [GitHub to submit the PR](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests).  And please be aware of [Zephyr's Contribution Guidelines](https://docs.zephyrproject.org/latest/contribute/index.html) before submitting PRs.  

Given the good feedback you have already shared, it would be great to include you as a contributor and benefit from your improvements.  Best regards

**Reply by rjmaris** [reply]
I just wrote a patch - see https://github.com/zephyrproject-rtos/zephyr/issues/59221
Related to forks: I know that it is easy, but it also eats server disk space.
